### PR TITLE
Visualize Optuna intermediate values and pruned trials

### PR DESCRIPTION
**Pull request by tonypzy**
## Description

- `py/visdom/integrations/optuna.py`
  - Record `FrozenTrial.intermediate_values` in training-step order.
  - Store reported values using the `intermediate_value` metric.
  - Preserve intermediate values for pruned trials.
  - Add an `optuna-intermediate-values` dashboard pane.
  - Skip the pane when no intermediate data is available or Optuna cannot generate the figure.

- `README.md`
  - Document intermediate-value recording and visualization.

## Motivation and Context

The Optuna integration covers trial logging, study dashboards and intermediate trial reporting.
The implementation is split into three smaller PRs:
- Optuna trial callback
- Study dashboard
- Intermediate trial reporting

This is the third PR and adds intermediate trial reporting.

## How Has This Been Tested?

Run the following code
```
from __future__ import annotations

import argparse
import math
import time
from collections import Counter

import numpy as np
import optuna
import visdom

from visdom.integrations import OptunaCallback


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Test Visdom Optuna intermediate reporting and pruning."
    )
    parser.add_argument("--server", default="http://localhost")
    parser.add_argument("--port", type=int, default=8097)
    parser.add_argument("--base-url", default="/")
    parser.add_argument(
        "--env",
        help="Dashboard environment; defaults to a unique timestamped name.",
    )
    parser.add_argument("--study-name", default="intermediate-pruning-demo")
    parser.add_argument("--n-trials", type=int, default=8)
    parser.add_argument("--steps", type=int, default=12)
    parser.add_argument("--refresh-every", type=int, default=2)
    parser.add_argument("--seed", type=int, default=7)
    args = parser.parse_args()
    if args.n_trials < 2:
        parser.error("--n-trials must be at least 2")
    if args.steps < 3:
        parser.error("--steps must be at least 3")
    if args.refresh_every < 1:
        parser.error("--refresh-every must be at least 1")
    return args


def dashboard_url(viz: visdom.Visdom, env: str) -> str:
    root = "{}:{}{}".format(viz.server, viz.port, viz.base_url).rstrip("/")
    return "{}/env/{}".format(root, env)


def main() -> None:
    args = parse_args()
    dashboard_env = args.env or "optuna_intermediate_pruning_{}".format(
        int(time.time())
    )
    viz = visdom.Visdom(
        server=args.server,
        port=args.port,
        base_url=args.base_url,
        env=dashboard_env,
        use_incoming_socket=False,
        raise_exceptions=True,
    )
    if not viz.check_connection(timeout_seconds=3):
        raise SystemExit(
            "Cannot connect to Visdom at {}:{}. Start the server first with "
            "`python3 -m visdom.server`.".format(args.server, args.port)
        )

    callback = OptunaCallback(
        viz,
        dashboard_env=dashboard_env,
        objective_names=["loss"],
        create_dashboard=True,
        refresh_every=args.refresh_every,
        raise_on_error=True,
    )
    study = optuna.create_study(
        study_name=args.study_name,
        direction="minimize",
        sampler=optuna.samplers.TPESampler(seed=args.seed),
        pruner=optuna.pruners.ThresholdPruner(
            upper=0.8,
            n_warmup_steps=2,
        ),
    )

    # These first two trials make the test deterministic: "good" completes and
    # "bad" is pruned once the pruner's warm-up period has elapsed.
    study.enqueue_trial({"profile": "good", "learning_rate": 0.01})
    study.enqueue_trial({"profile": "bad", "learning_rate": 0.01})

    def objective(trial: optuna.Trial) -> float:
        profile = trial.suggest_categorical("profile", ["good", "medium", "bad"])
        learning_rate = trial.suggest_float("learning_rate", 1e-4, 1e-1, log=True)
        initial_loss = {"good": 0.55, "medium": 0.95, "bad": 1.5}[profile]
        decay = {"good": 0.84, "medium": 0.93, "bad": 0.98}[profile]
        learning_rate_penalty = 0.04 * abs(math.log10(learning_rate) + 2.0)
        trial_env = callback.trial_env(trial)
        loss = initial_loss

        for step in range(args.steps):
            loss = initial_loss * decay**step + learning_rate_penalty
            trial.report(loss, step=step)
            viz.line(
                X=np.asarray([step], dtype=np.float64),
                Y=np.asarray([loss], dtype=np.float64),
                win="live-training-loss",
                env=trial_env,
                update="append",
                opts={
                    "title": "Live Training Loss",
                    "xlabel": "step",
                    "ylabel": "loss",
                },
            )
            if trial.should_prune():
                raise optuna.TrialPruned(
                    "loss {:.4f} exceeded the pruning threshold at step {}".format(
                        loss, step
                    )
                )

        return loss

    study.optimize(objective, n_trials=args.n_trials, callbacks=[callback])
    if not callback.update_dashboard(study):
        raise RuntimeError("the final Optuna dashboard update failed")

    states = Counter(trial.state.name for trial in study.trials)
    if not states["COMPLETE"] or not states["PRUNED"]:
        raise AssertionError(
            "expected both COMPLETE and PRUNED trials, got {}".format(dict(states))
        )
    if any(not trial.intermediate_values for trial in study.trials):
        raise AssertionError("every trial should contain intermediate values")

    print("\nOptuna intermediate reporting and pruning test passed.")
    print("Trial states: {}".format(dict(states)))
    print("Dashboard environment: {}".format(dashboard_env))
    print("Dashboard URL: {}".format(dashboard_url(viz, dashboard_env)))


if __name__ == "__main__":
    main()
```

## Screenshots (if appropriate):

https://github.com/user-attachments/assets/0e99a3a2-a8f2-48f1-a668-b40f05e44f32

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Add intermediate-value tracking and optional dashboard visualization to the Optuna integration.

New Features:
- Record Optuna intermediate trial values as ordered Visdom metrics, including values from pruned trials.
- Add optional Optuna study dashboards with summary, trial, optimization, timeline, parameter-importance, and intermediate-value visualizations.

Enhancements:
- Support configurable dashboard creation and periodic refreshes with safe handling when visualization dependencies or data are unavailable.
- Link dashboard summaries to the best and latest trial environments.

Documentation:
- Document Optuna intermediate-value recording, dashboard usage, and Plotly requirements.

## Summary by CodeRabbit

* **New Features**
  * Added optional Optuna study dashboards with configurable refresh intervals.
  * Dashboards include study summaries, hyperparameters, optimization history, parameter importance, intermediate metrics, timelines, and links to individual trials.
  * Added safer trial navigation and support for studies with or without named metrics.
  * Dashboard updates can warn or raise errors based on configuration.
  * Studies without completed or feasible trials no longer fail when displaying best-trial information.

* **Documentation**
  * Expanded Optuna integration guidance, including Plotly setup, dashboard configuration, supported trial states, panes, metrics, links, and post-trial recording behavior.

### CODE REVIEW

**Comment by tonypzy** [comment]
@vedansh-5 @rajnisht7 @Manik-Khajuria-5 @Jayantparashar10 Please review

File: py/visdom/integrations/optuna.py [hunk 1]
```diff
@@ -163,6 +197,235 @@ def _trial_tags(self, study: Any, trial: Any) -> dict[str, str]:
...
+
+    def update_dashboard(self, study: Any) -> bool:
+        """Create or refresh all dashboard panes for ``study``.
+
+        Returns whether the dashboard was written successfully. Only trials
+        logged by this callback instance are included in the HParams pane;
+        loading trials from a resumed study is handled by the later resume
+        integration.
+        """
+        payload = self._build_dashboard_payload(study)
```
Comment (Manik-Khajuria-5) [inline py/visdom/integrations/optuna.py]: @tonypzy I think this must be inside try catch block

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy Please look into below mentioned code changes

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy Also add try catch in build Dashboard function and also I think trial env must be escaped env

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy also right now are accepting values that may result in invalid refresh behavior. It would be safer to reject non positive values in the refresh logic. Please make the changes wherever neccessary

**Comment by tonypzy** [comment]
> @tonypzy also right now are accepting values that may result in invalid refresh behavior. It would be safer to reject non positive values in the refresh logic. Please make the changes wherever neccessary

please see the latest commit

**Review by Jayantparashar10 — COMMENTED** [review]
@tonypzy Please resolve conflicts

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy Could we use getattr(study, metric names, None) here? This would keep the integration compatible with older Optuna versions where metric names isn't available.

**Review by rajnisht7 — APPROVED** [review]

**Review by Jayantparashar10 — APPROVED** [review]

**Comment by Jayantparashar10** [comment]
@tonypzy please resolve conflicts

**Review by Manik-Khajuria-5 — APPROVED** [review]
@tonypzy LGTM!
